Sources/Datadog/DatadogInternal/Context/DateCorrector.swift

```diff
@@ -0,0 +1,13 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+
+/// Calculates the date correction for adjusting device time to server time.
+internal protocol DateCorrector {
+    /// Returns recent date correction for adjusting device time to server time.
+    var offset: TimeInterval { get }
+}
```

Comment:
It is no longer a "date corrector" IMO if we make it return `offset`. The (previous) idea of having `DateCorrector` was to return a `DateCorrection` type that makes it impossible to make mistake:
```swift
internal struct DateCorrection {
   // ...

   func applying(to deviceDate: Date) -> Date {
      return deviceDate.addingTimeInterval(serverTimeOffset)
   }
}
```
Now if we want to work with `offset`, an instant question is _"should we add it or subtract it?"_. This IMO sounds very error-prone. I'd prefer having type-safety on date correction, instead of passing arbitrary `TimeInterval` around, like in RUM scopes:

```diff
- date: dateCorrection.applying(to: resourceStartTime).timeIntervalSince1970.toInt64Milliseconds,
+ date: resourceStartTime.addingTimeInterval(serverTimeOffset).timeIntervalSince1970.toInt64Milliseconds,
```

@maxep What is the motivation behind dropping `DateCorrection` 🤔💭?

Reply:
Arrg my previous comment didn't go through! sorry about that:

This is arguable 🙂 I don't see the value of this actually, there is no sign manipulation. IMO it is more error-prone to not apply the offset than to wrongly apply it by inverting the sign, it is also super easy to use the wrong reference date (`timeIntervalSince1970` vs. `timeIntervalSinceReferenceDate`) or to use the wrong conversion (`toInt64Milliseconds` vs. `toInt64Nanoseconds`). So even now, we would always rely on either doc (which I should add) or previous usage to build the correct timestamp for an event.

The motivation is to reduce the number of types, we have many custom types for date manipulation that it gets confusing. I would definitely agree if this type was public and used externally, but this is for internal usage.

Now, we could have a `Date` extension for internal usage. By leveraging [this pattern](https://github.com/Alamofire/Alamofire/blob/master/Source/AlamofireExtended.swift), we could do something like:
```swift
// Date extension method:
func timestamp(byApplying offset: TimeInterval) -> Int64 {
    return self.addingTimeInterval(serverTimeOffset).timeIntervalSince1970.toInt64Milliseconds
}

date.dd.timestamp(byApplying: offset)
```

IMO, the properties need to be documented, either if it's a `DateCorrection` or a `TimeInterval`. In both cases, it requires the same effort to build the timestamp.

Reply:
Indeed there are other transformations we do (e.g. `timeIntervalSince1970` and `toInt64Milliseconds`) worth solidifying - I didn't look that far. Single extension is debatable as we use different units in RUM (milliseconds) and Tracing (nanoseconds). Maybe a separate extension in each Feature?

Anyway, I'm ok with `serverTimeOffset` to move forward and let's keep this further refactoring in backlog.

Reply:
> RUM (milliseconds) and Tracing (nanoseconds)

I didn't even noticed 😄